File: 卡码/代码随想录/C++/动态规划/dp/05-子序列问题/src/04-回文/LeetCode_0647.cpp
//647. 回文子串
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
	//动态规划
	int countSubstrings(string s) {
		vector<vector<bool>> dp(s.length(),vector<bool>(s.length(),false));
		int res=0;
		for(int i=s.length()-1;i>=0;i--){
			for(int j=i;j<s.length();j++){
				if(s[i]==s[j]&&(j-i<=1||dp[i+1][j-1])){
					dp[i][j]=true;
					res++;
				}
			}
		}
		return res;
	}
	//双指针
	int countSubstrings1(string s) {
		int res=0;
		for(int i=0;i<s.length();i++){
			res+=countPalindrome(s,i,i);
			res+=countPalindrome(s,i,i+1);
		}
		return res;
	}
	int countPalindrome(const string& s,int i,int j){
		int res=0;
		while(i>=0&&j<s.length()&&s[i]==s[j]){
			i--;
			j++;
			res++;
		}
		return res;
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/1/09-动态规划/Java/dp/src/main/java/com/kurumi/dp/_05_子序列问题/_04_回文/LeetCode_0647.java
//647. 回文子串
package com.kurumi.dp._05_子序列问题._04_回文;

public class LeetCode_0647 {

}
class Solution {
  //动态规划
  public int countSubstrings(String s) {
    int n=s.length();
    boolean[][] dp=new boolean[n][n];
    int res=0;
    for(int i=n-1;i>=0;i--){
      for(int j=i;j<n;j++){
        if(s.charAt(i)==s.charAt(j)&&(j-i<=1||dp[i+1][j-1]==true)){
          dp[i][j]=true;
          res++;
        }
      }
    }
    return res;
  }
  //双指针
  public int countSubstrings1(String s) {
    int res=0;
    for(int i=0;i<s.length();i++){
      res+=countPalindrome(s, i, i);
      res+=countPalindrome(s, i, i+1);
    }
    return res;
  }
  public int countPalindrome(String s,int i,int j){
    int res=0;
    while(i>=0&&j<s.length()&&s.charAt(i)==s.charAt(j)){
      i--;
      j++;
      res++;
    }
    return res;
  }
}
